File: ch01/python/ollama-local-basic/ollama-local-basic.py
"""
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import os
import time

import requests

OLLAMA_HOST = os.getenv("OLLAMA_HOST", "http://localhost:11434")
OLLAMA_MODEL = os.getenv("OLLAMA_MODEL", "gemma3:4b")


def query_model(prompt: str,
                model: str = OLLAMA_MODEL,
                temperature: float = 0) -> str:
    """Send a user prompt to a local LLM running through Ollama."""
    payload = {
        "model": model,
        "prompt": prompt,
        "stream": False,
        "options": {
            "temperature": temperature,
        },
    }

    start = time.perf_counter()
    response = requests.post(f"{OLLAMA_HOST}/api/generate",
                             json=payload,
                             timeout=120)
    # ...
    latency = time.perf_counter() - start
    response.raise_for_status()
    result = response.json()

    input_tokens = result.get("prompt_eval_count", 0)
    output_tokens = result.get("eval_count", 0)
    print(f"\tModel: {result.get('model', model)}")
    print(f"\tLatency: {latency:.3f} seconds")
    print(f"\tInput tokens: {input_tokens}")
    print(f"\tOutput tokens: {output_tokens}")
    print(f"\tTotal tokens: {input_tokens + output_tokens}")

    return result["response"].strip()


if __name__ == "__main__":
    prompt = "How many tokens is your context window?"

    print("User:", prompt)
    response = query_model(prompt)
    print("Local LLM:", response)

File: ch01/javascript/ollama-local-basic/ollama-local-basic.js
/*
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { performance } from 'node:perf_hooks';

const ollamaHost = process.env.OLLAMA_HOST || 'http://localhost:11434';
const ollamaModel = process.env.OLLAMA_MODEL || 'gemma3:4b';

async function queryModel(userPrompt, model = ollamaModel, temperature = 0) {
    const payload = {
        model: model,
        prompt: userPrompt,
        stream: false,
        options: {
            temperature: temperature,
        },
    };

    const start = performance.now();
    const response = await fetch(`${ollamaHost}/api/generate`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    });
    const latency = (performance.now() - start) / 1000;

    if (!response.ok) {
        throw new Error(`Ollama request failed: ${await response.text()}`);
    }

    const result = await response.json();
    const inputTokens = result.prompt_eval_count || 0;
    const outputTokens = result.eval_count || 0;

    console.log(`\tModel: ${result.model || model}`);
    console.log(`\tLatency: ${latency.toFixed(3)} seconds`);
    console.log(`\tInput tokens: ${inputTokens}`);
    console.log(`\tOutput tokens: ${outputTokens}`);
    console.log(`\tTotal tokens: ${inputTokens + outputTokens}`);

    return result.response.trim();
}

const userPrompt = 'How many tokens is your context window?';

console.log('User:', userPrompt);
const response = await queryModel(userPrompt);
console.log('Local LLM:', response);
